Assets/MixedRealityToolkit.SDK/Experimental/Features/UX/MixedRealityKeyboardBase.cs

```diff
@@ -0,0 +1,160 @@
+
+        private Coroutine stateUpdate;
+
+        #endregion private fields
+
+        #region Private enums
+
+        private enum KeyboardState
+        {
+            Hiding,
+            Hidden,
+            Showing,
+        }
+
+        #endregion Private enums
+
+        #region Unity functions
+
+        private void Awake()
+        {
+            if ((inputField = GetComponent<T>()) == null)
+            {
+                Destroy(this);
+            }
+        }
+
+        private void Start()
+        {
+            UnityEngine.WSA.Application.InvokeOnUIThread(() =>
+            {
+                inputPane = InputPane.GetForCurrentView();
+                inputPane.Hiding += (inputPane, args) => OnKeyboardHiding();
+                inputPane.Showing += (inputPane, args) => OnKeyboardShowing();
+            }, false);
+        }
+
+        private IEnumerator UpdateState()
+        {
+            switch (State)
+            {
+                case KeyboardState.Showing:
+                    UpdateText(keyboard?.text);
+                    break;
+
+                case KeyboardState.Hiding:
+                    ClearText();
+                    State = KeyboardState.Hidden;
+                    break;
+            }
+
+            yield return null;
+        }
+
+        private void OnDisable()
+        {
+            HideKeyboard();
+        }
+
+        #endregion unity functions
+
+        protected void HideKeyboard()
+        {
+            ClearText();
+            State = KeyboardState.Hidden;
+            UnityEngine.WSA.Application.InvokeOnUIThread(() => inputPane?.TryHide(), false);
+
+            if (stateUpdate != null)
+            {
+                StopCoroutine(stateUpdate);
+                stateUpdate = null;
+            }
+        }
+
+        protected void ShowKeyboard()
+        {
+            // 2019/08/14: We show the keyboard even when the keyboard is already visible because on HoloLens 1
+            // and WMR the events OnKeyboardShowing and OnKeyboardHiding do not fire
+            //if (state == KeyboardState.Showing)
+            //{
+            //    Debug.Log($"MixedRealityKeyboard.ShowKeyboard called but keyboard already visible.");
+            //    return;
+            //}
+
+            State = KeyboardState.Showing;
+
+            if (keyboard != null)
+            {
+                keyboard.text = string.Empty;
+                UnityEngine.WSA.Application.InvokeOnUIThread(() => inputPane?.TryShow(), false);
+            }
+            else
+            {
+                keyboard = TouchScreenKeyboard.Open("", TouchScreenKeyboardType.Default, false, false, false, false);
+            }
+
+            if (stateUpdate == null)
+            {
+                stateUpdate = StartCoroutine(UpdateState());
+            }
+        }
+
+        #region Input pane event handlers
+
+        private void OnKeyboardHiding()
+        {
+            if (State != KeyboardState.Hidden)
```

Comment:
seems weird OnHiding we do state check/change but not on showing?

Reply:
same as above, code copy

Reply:
I guess once the keyboard related bugs are fixed it's much easier to write this class